Generate blackbox targets for snmp_exporter service for all sites.

nkinkade opened the pull request:
Then Prometheus snmp_exporter service is currently deployed in a GCE instance in the mlab-sandbox project. It will later also be deployed to mlab-staging and mlab-oti. This PR updates mlabconfig.py to generate Prometheus blackbox target configurations for the snmp_exporter service.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/m-lab/operator/152)
<!-- Reviewable:end -->

stephen-soltesz commented:
Hey, Nathan,

Sorry for the delays here. I've added comments. I think the only issue is around using `snmp-targets` like we discussed last week and one other minor name choice.

---

Reviewed 1 of 3 files at r1.
Review status: 1 of 3 files reviewed at latest revision, 3 unresolved discussions.

---

*[generate_prometheus_targets.sh, line 10 at r1](https://reviewable.io:443/reviews/m-lab/operator/152#-KuvLpMgod0pzPTKbvHj:-KuvLpMgod0pzPTKbvHk:buuprgu) ([raw file](https://github.com/m-lab/operator/blob/e680dd813261e30053a1914073f3502963ebf61a/generate_prometheus_targets.sh#L10)):*
> ```Shell
> # Create all output directories.
> for project in mlab-sandbox mlab-staging mlab-oti ; do
>   mkdir -p ${BASEDIR}/gen/${project}/prometheus/{legacy-targets,blackbox-targets}
> ```

+`snmp-targets` (or similar).

---

*[generate_prometheus_targets.sh, line 49 at r1](https://reviewable.io:443/reviews/m-lab/operator/152#-KuvHSYAdbREgClepCvS:-KuvHSYAdbREgClepCvT:bui26im) ([raw file](https://github.com/m-lab/operator/blob/e680dd813261e30053a1914073f3502963ebf61a/generate_prometheus_targets.sh#L49)):*
> ```Shell
>         --template_target=s1.{{sitename}}.measurement-lab.org:9116 \
>         --label service=snmp_exporter > \
>         ${output}/blackbox-targets/snmpexporter.json
> ```

Since these configs will be for the snmp exporter, I suggest using an output directory like `${output}/snmp-targets/snmpexporter.json`.

Ultimately, the prometheus snmp exporter config will reference files in a subdirectory like this:
https://github.com/m-lab/prometheus-support/blob/master/config/federation/prometheus/prometheus.yml#L334

Prometheus makes it possible to enumerate the files explicitly, but using patterns allows adding new configs ad-hoc quickly without reloading the server. Given the number of separate directories we have now, it's becoming a little cumbersome, so push back if you think a different way would be more helpful than maintaining separate directories for different configs.

---

*[plsync/mlabconfig.py, line 715 at r1](https://reviewable.io:443/reviews/m-lab/operator/152#-KuvIFbg59H8l3br_Bvt:-KuvIFbg59H8l3br_Bvu:b-c0dzaj) ([raw file](https://github.com/m-lab/operator/blob/e680dd813261e30053a1914073f3502963ebf61a/plsync/mlabconfig.py#L715)):*
> ```Python
>             continue
>         labels = common_labels.copy()
>         labels['module'] = site['name']
> ```

Part of me wants to keep this generic. And, iiuc, 'module' as a name is very particular to the use-case for the snmp exporter.  What do you think of calling this label name 'site' instead? We'll be able to reference and set the 'site' value for the 'module' parameter for the snmp exporter in the prometheus relabel config later.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/m-lab/operator/152#-:-KuvLgTpi1gwOsMzCAPj:bbzzf6a)*
<!-- Sent from Reviewable.io -->

nkinkade commented:
@stephen-soltesz: I think I have addressed your questions. PTAL?  Thanks!

---

Review status: 1 of 3 files reviewed at latest revision, 3 unresolved discussions.

---

*[generate_prometheus_targets.sh, line 49 at r1](https://reviewable.io:443/reviews/m-lab/operator/152#-KuvHSYAdbREgClepCvS:-Kuzyem6KhY_8gEXJfby:bvf5w8i) ([raw file](https://github.com/m-lab/operator/blob/e680dd813261e30053a1914073f3502963ebf61a/generate_prometheus_targets.sh#L49)):*
<details><summary><i>Previously, stephen-soltesz (Stephen Soltesz) wrote…</i></summary><blockquote>

Since these configs will be for the snmp exporter, I suggest using an output directory like `${output}/snmp-targets/snmpexporter.json`.

Ultimately, the prometheus snmp exporter config will reference files in a subdirectory like this:
https://github.com/m-lab/prometheus-support/blob/master/config/federation/prometheus/prometheus.yml#L334

Prometheus makes it possible to enumerate the files explicitly, but using patterns allows adding new configs ad-hoc quickly without reloading the server. Given the number of separate directories we have now, it's becoming a little cumbersome, so push back if you think a different way would be more helpful than maintaining separate directories for different configs.
</blockquote></details>

I've added a new `snmp-targets` directory.

There are currently only [blackbox-targets and legacy-targets](https://console.cloud.google.com/storage/browser/operator-mlab-sandbox/prometheus/?project=mlab-sandbox), right? Or are you referring to all the mount point in the Prom cluster (e.g., aeflex-targets, federation-targets, legacy-targets, etc.)?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/m-lab/operator/152#-:-KuzzlqaoB_zuEbP5Yzc:b4nlj88)*
<!-- Sent from Reviewable.io -->

stephen-soltesz commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> Thanks!

---

Reviewed 2 of 2 files at r2.
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[generate_prometheus_targets.sh, line 49 at r1](https://reviewable.io:443/reviews/m-lab/operator/152#-KuvHSYAdbREgClepCvS:-Kv-IRsdlQQnqBtteE7S:b1bj10r) ([raw file](https://github.com/m-lab/operator/blob/e680dd813261e30053a1914073f3502963ebf61a/generate_prometheus_targets.sh#L49)):*
<details><summary><i>Previously, nkinkade wrote…</i></summary><blockquote>

I've added a new `snmp-targets` directory.

There are currently only [blackbox-targets and legacy-targets](https://console.cloud.google.com/storage/browser/operator-mlab-sandbox/prometheus/?project=mlab-sandbox), right? Or are you referring to all the mount point in the Prom cluster (e.g., aeflex-targets, federation-targets, legacy-targets, etc.)?
</blockquote></details>

I was referring to all the mount points for the Prometheus deployment. This will be okay for now.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/m-lab/operator/152#-:-Kv-J48JsGnCuje0OogM:b-g2a5fb)*
<!-- Sent from Reviewable.io -->

stephen-soltesz commented:
Oh, it looks like the test needs to be updated for the s/module/site change.

nkinkade commented:
Fixed unit test, and Travis CI build now passing. Merging.
